syntax="proto3";

import "common.proto";

package session;

message SessionMsg {                                // 奇数是请求，偶数是响应
    oneof choice {
        common.TestReq      test_req    = 1;        // 请求
        common.TestRsp      test_rsp    = 2;        // 响应

        CreateSessionReq    create_session_req  = 3;    // 创建会话  
        CreateSessionRsp    create_session_rsp  = 4;

        DelSessionReq       del_session_req     = 5;    // 删除会话
        DelSessionRsp       del_session_rsp     = 6;

        QuerySessionReq     query_session_req   = 7;    // 查询会话
        QuerySessionRsp     query_session_rsp   = 8;

        QuerySessionListReq query_session_list_req  = 9;  // 批量查询会话
        QuerySessionListRsp query_session_list_rsp  = 10;
    }
}

enum ErrorCode {
    SUCCESS                         = 0;

    // 系统错误码 (各个服务service_id * 10000 + 公共系统错误码)
    ERR_SYS_BEGIN                   = 104000100;
    ERR_SYS_OVERLOAD                = 104000101;    // 服务过载
    ERR_SYS_REJECT_SERVICE          = 104000102;    // 拒绝服务
    ERR_SYS_SERVER_INNER            = 104000103;    // 服务内部错误
    ERR_SYS_TIMEOUT                 = 104000104;    // 超时错误
    ERR_SYS_NO_INSERVICE_LIST       = 104000105;    // 没有可用服务ip
    ERR_SYS_TASK_STATE              = 104000106;    // 任务状态错误
    ERR_SYS_TASK_DISCARD            = 104000107;    // 任务丢弃
    ERR_SYS_END                     = 104000199;

    // 数据包错误码 (各个服务service_id * 10000 + 公共数据包错误码)
    ERR_PACKET_BEGIN                = 104000200;
    ERR_PACKET_ENCODE               = 104000201;   	// 打包失败
    ERR_PACKET_DECODE               = 104000202;   	// 解码失败
    ERR_PACKET_HEADER               = 104000203;
    ERR_PACKET_LEN                  = 104000204;
    ERR_PACKET_VERSION              = 104000205;
    ERR_PACKET_FROM_SERVICE_ID      = 104000206;
    ERR_PACKET_TO_SERVICE_ID        = 104000207;
    ERR_PACKET_APP_ID               = 104000208;
    ERR_PACKET_APP_VERSION          = 104000209;
    ERR_PACKET_CONN_SEQ_ID          = 104000210;
    ERR_PACKET_MSG_SEQ_ID           = 104000211;
    ERR_PACKET_DATA_FORMAT          = 104000212;
    ERR_PACKET_DATA_FIELD_0         = 104000213;
    ERR_PACKET_DATA_FIELD_1         = 104000214;
    ERR_PACKET_DATA_FIELD_2         = 104000215;
    ERR_PACKET_DATA_FIELD_3         = 104000216;
    ERR_PACKET_CHECK_SUM            = 104000217;
    ERR_PACKET_UNKNOWN_REQUEST      = 104000218;   	// 不明请求
    ERR_PACKET_END                  = 104000299;

    // 接口错误码 (各个服务service_id * 10000 + 公共接口错误码)
    ERR_INTERFACE_BEGIN             = 104000300;
    ERR_INTERFACE_PARAM             = 104000301;   	// 参数错误
    ERR_INTERFACE_PERM              = 104000302;   	// 权限错误
    ERR_INTERFACE_TIMEOUT           = 104000303;   	// 接口超时
    ERR_INTERFACE_MASTER_TIMEOUT    = 104000304;    // 接口主任务超时
    ERR_INTERFACE_SUB_TIMEOUT       = 104000305;    // 接口子任务超时
    ERR_INTERFACE_END               = 104000399;

    // 业务错误码 (各个服务service_id * 10000 + 公共业务错误码)
    ERR_BUSINESS_BEGIN              = 104001000;
    ERR_BUSINESS_END                = 104009999;
}

message Session {
    uint64  user_id             = 1;
    int32   service_id          = 2;    // 方便gate就近接入使用,不同地区gate的service_id不同
    uint32  proc_id             = 3;
    bytes   in_ip               = 4;
    int32   in_port             = 5;
    uint64  conn_id             = 6;    // conn_id为0，表示离线
    uint32  create_time         = 7;
    int32   version             = 8;
    uint32  app_id              = 9;
    uint32  app_version         = 10;
    uint32  dev_type            = 11;    
}

message ConnKey {
    bytes   in_ip       = 1;
    int32   in_port     = 2;
    uint64  conn_id     = 3;
}

message CreateSessionReq {
    Session     session     = 1;
}

message CreateSessionRsp {

}

message DelSessionReq {
    uint64  user_id         = 1;    // user_id优先(user_id为0时，用conn_key删除)
    ConnKey conn_key        = 2;
}

message DelSessionRsp {

}

message QuerySessionReq {
    uint64  user_id     = 1;
 }

message QuerySessionRsp {
    Session session     = 1;
}

message QuerySessionListReq {
    repeated uint64     user_id_list    = 1;
 }

message QuerySessionListRsp {
    repeated Session    session_list    = 1;
}


